alphagov/pay-publicapi
# PP-5285 Convert swagger to Openapi v3

## WHAT YOU DID
- Convert swagger to Openapi v3
- Removed swagger-maven-plugin (groupId: com.github.kongchen)

## How to test
- Npm packages `swagger2openapi` (to convert current swagger to openapi) & `json-diff` can be used to find differences with previous version and v3 (current version). 
[Click here to see differences or run commmands below ](https://github.com/alphagov/pay-publicapi/files/3695846/swaggerv2_v3_diff.pdf)

```
npm install swagger2openapi
./node_modules/swagger2openapi/swagger2openapi.js -o swagger_old.json https://raw.githubusercontent.com/alphagov/pay-publicapi/ab938d0f79470566fb63bd6638de44c6e6b0ffd6/swagger/swagger.json

wget --output-document=swagger_new.json https://raw.githubusercontent.com/alphagov/pay-publicapi/599452aac435e9e0b8215aaa3ba089495d281d96
[swaggerv2_v3_diff.pdf](https://github.com/alphagov/pay-publicapi/files/3695835/swaggerv2_v3_diff.pdf)
/swagger/swagger.json

npm install json-diff
./node_modules/json-diff/bin/json-diff.js swagger_old.json swagger_new.json
```


Below are major set of differences due to improvements
- example: added examples for attributes
- required attribute: `required=false` has been removed at many places as default is false
- security: security definition has been upto BearerAuth
- readOnly: `readOnly=true` has been removed for `Create` schemas as data is not read only ex: CreateCardPaymentRequest, CreateDirectDebitPaymentRequest
- Metadata: updated from object type to schema `ExternalMetadata`
